### PR TITLE
[Reformer classification head] Implement the reformer model classification head for text classification

### PR DESCRIPTION
**Pull request by as-stevens**
This PR is an effort to implement the classification head on top of the plain vanilla reformer model.
References have been taken from the Roberta and XLNet classification head to implement the reformer classification head.

Tried testing the implementation changes, but it is failing error, link to google collab https://colab.research.google.com/drive/1KFsQxLqsMB6vBF4_bRmTFGhdGwkgx0zI?usp=sharing;
Test scenario details;

- IMDB movie review dataset.

-  The cell3 has the reformer model classification head code.

- Since the length of the reviews is not more than 512 words( assuming there would be a few outliers, but that could be skipped)

- Initializing the model with the config.axial_pos_shape to (16,32) but the setting does not take effect and throws runtime error.

Please help.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Looks cool :-) Will take a look soon!

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1832]: Could you replace all `x` and `features` with `hidden_states`? 

1) It would be great to keep reformer as memory efficient as possible -> only use one variable name in the function.
2) Even though the variable name `x` is used in other models, it's bad practice and we try to discontinue this and use more explicit variable names (also better for serach and replace)

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1827]: Here we have to be careful in Reformer because it uses reversible revnets. This means that self.dence should be `nn.Linear(2 * config.hidden_size, config.hidden_size)`

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1889]: `do_output_hidden_states` is renamed to `output_hidden_states` 
and we should also add `output_attentions` here.

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1874]: I don't think we need this function `get_output_embeddings()` here. `self.lm_head` also does not exist.

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1861]: Not needed -> can delete this line

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1862]: Not needed -> can delete this line

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py:1827]: Noted, will make the necessary change

**Comment by patrickvonplaten** [comment]
I left a couple of comments that colud help you fix the PR :-) Also it would be great to add tests here

**Comment by as-stevens** [comment]
> I left a couple of comments that colud help you fix the PR :-) Also it would be great to add tests here

Could you please point to some reference code base, that I can look at. This will give me a good idea of the testing standard in the application.

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py:1832]: Noted

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py:1861]: Noted

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py:1862]: Noted

**Comment by patrickvonplaten** [comment]
Sure, here we go: https://github.com/huggingface/transformers/blob/1ae132a07d7f294cf58cd50f7db8723d00e282de/tests/test_modeling_bert.py#L361

and 

https://github.com/huggingface/transformers/blob/1ae132a07d7f294cf58cd50f7db8723d00e282de/tests/test_modeling_bert.py#L520

All this class should be added to `all_model_classes` in the Reformer test.

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py]: As mentioned removed the get_output_embeddings()

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py]: changed the variables names and mentioned

File: tests/test_modeling_reformer.py
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:986]: This makes only sense if the model in question is a sequence classification model and `google/reformer-enwik8` is not. Would be nice to just add one "normal" (non-slow) test as is done in Bert.

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1865]: Also, it would be great to add a warning here:

```python 
if config.is_decoder is True:
      logging.warning("You might want to disable causal masking for sequence classification")
```

**Comment by as-stevens** [comment]
@patrickvonplaten I tried to fix the test case for the reformer classification head. But it is failing in one of the common test cases class,  
test_modelling_common.py-> test_hidden_states_output

where it tries to run the hidden states test for the Reformer Models, it fails only for classification head.
Since this is a common workflow, I am wondering where am I going wrong. Also, the newly added QA head on the reformer is not failing which is quite surprising. It looks like it bypasses the QA head and skips the test.

Please let me know your thoughts/suggestions.
Thanks
Amit

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1857]: `output_hidden_states=None` 
and 
`output_attentions=None` 
here

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1868]: `output_attentions=output_attentions` is missing here.

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [reply src/transformers/modeling_reformer.py:1857]: The default is **always** None

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [reply src/transformers/modeling_reformer.py:1868]: This will probably fix the test.

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py]: Fixed as suggested

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py:1854]: Thanks a ton! this works and test cases are fixed.

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1846]: Please add docstring here similar to Bert:
```python
        r"""
        labels (:obj:`torch.LongTensor` of shape :obj:`(batch_size,)`, `optional`, defaults to :obj:`None`):
            Labels for computing the sequence classification/regression loss.
            Indices should be in :obj:`[0, ..., config.num_labels - 1]`.
            If :obj:`config.num_labels == 1` a regression loss is computed (Mean-Square loss),
            If :obj:`config.num_labels > 1` a classification loss is computed (Cross-Entropy).
        """
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1871]: Since very recently we have something similar to namedtuples => could you apply this here as well? 
Should rather straight-forward if you check out https://github.com/huggingface/transformers/blob/ce374ba87767d551f720242d5e64bfa976531079/src/transformers/modeling_bert.py#L1298

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1834]: Missing docstring:

```python 
    @add_start_docstrings_to_callable(REFORMER_INPUTS_DOCSTRING)
    @add_code_sample_docstrings(
        tokenizer_class=_TOKENIZER_FOR_DOC,
        checkpoint="google/reformer-crime-and-punishment"
        output_type=SequenceClassifierOutput,
        config_class=_CONFIG_FOR_DOC,
    )
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1819]: For consistency, could you move the whole class after `ReformerForMaskedLM` ?

**Comment by patrickvonplaten** [comment]
I think we are very close to merging this PR :-) After the minor changes as described above, I think we are good to merge

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py]: Moved the classes after the ReformerForMaskedLM.

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py]: Added the docstring to the reformer classification head

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py]: Implemented as suggested.

File: src/transformers/modeling_reformer.py
Comment (as-stevens) [reply src/transformers/modeling_reformer.py]: Added the docstring.

**Comment by as-stevens** [comment]
@patrickvonplaten Implemented the changes as suggested :)

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1788]: Think we can remove this docstring as its already defined above :-)

**Comment by patrickvonplaten** [comment]
Great, can you just remove the duplicated docstring `REFORMER_START_DOCSTRING` ? After that we are good to merge :-)

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1845]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1788]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1789]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1790]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1791]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1792]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1793]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1794]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1795]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1796]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1797]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1798]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1799]:
```suggestion
```

File: src/transformers/modeling_reformer.py
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:1800]:
```suggestion
```

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Waiting for another approval - LGTM

File: src/transformers/modeling_reformer.py
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:1879]: I would be more specific about how they might disable causal masking. Or raise.

File: src/transformers/modeling_reformer.py
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:1939]: F.mse_loss, F.cross_entropy_loss?

File: src/transformers/modeling_reformer.py
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:1960]: `nn.ModuleList()` would be nice here

File: tests/test_modeling_reformer.py
Comment (sshleifer) [inline tests/test_modeling_reformer.py:456]: can all be one line. 
Can add coverage by saying 
```python
if torch_device == 'fp16': model.half()
```

File: tests/test_modeling_reformer.py
Comment (sshleifer) [inline tests/test_modeling_reformer.py:463]: not a fan of `self.check_loss_output` .
You can't see what it's asserting and it's asserting very little and it's taking a much larger amount of data than it uses.

File: tests/test_modeling_reformer.py
Comment (sshleifer) [reply tests/test_modeling_reformer.py:456]: is this not covered by common tests?

File: tests/test_modeling_reformer.py
Comment (sshleifer) [inline tests/test_modeling_reformer.py:602]: test_pruning = False
    test_headmasking = False
    test_torchscript = False


Obviously out of scope, but you could try turning some of these on and seeing whether/why they break.

**Review by sshleifer — APPROVED** [review]
sorry for being slow. Ping me!

**Comment by jstremme** [comment]
Awesome PR @as-stevens, do you have a new example notebook?  If not I take it that it's possible to just import the class you created in your notebook and have added to the codebase.

**Comment by as-stevens** [comment]
@jstremme the link to the notebook, where I tried to fine-tune the model for IMDB text classification. Since there are not many pre-trained models for Reformer, I used the Crime and Punishment for classification. 
[https://colab.research.google.com/drive/1l61NccWTGMfNFPj1T8kvMjnik2iEQWfe](url)

The model did not perform well, as it is not a bi-directional trained model. The link [https://github.com/huggingface/transformers/issues/5023](url) contains more information.

Thanks
Amit

**Comment by jstremme** [comment]
Thanks very much @as-stevens.  Great work, and I'm hopeful I'll be able to pretrain and use my own bi-directional Reformer for text classification by following a similar approach.  Happy to share high-level details with the transformers community as I go.

**Pull request merged**
